### PR TITLE
sparkplug: Telemetry

### PR DESCRIPTION
**Pull request by nickAS21**
## Pull Request description

- added DeviceTelemetryProto with Sparkplug (node, device)
- added SparkplugMetricUtil

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java [hunk 1]
```diff
@@ -324,15 +326,13 @@ private void processPublish(ChannelHandlerContext ctx, MqttPublishMessage mqttMs
         String topicName = mqttMsg.variableHeader().topicName();
         int msgId = mqttMsg.variableHeader().packetId();
         log.trace("[{}][{}] Processing publish msg [{}][{}]!", sessionId, deviceSessionCtx.getDeviceId(), topicName, msgId);
-
-        if (sparkplugSessionHandler != null) {
-            handleSparkplugPublishMsg(ctx, topicName, msgId, mqttMsg);
-            transportService.reportActivity(deviceSessionCtx.getSessionInfo());
-        } else if (topicName.startsWith(MqttTopics.BASE_GATEWAY_API_TOPIC)) {
+        if (topicName.startsWith(MqttTopics.BASE_GATEWAY_API_TOPIC)) {
             if (gatewaySessionHandler != null) {
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:330]: We should throw error if ```gatewaySessionHandler is null```

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+/**
+ * Provides utility methods for SparkplugB MQTT Payload Metric.
+ */
+@Slf4j
+public class SparkplugMetricUtil {
+
+    public static Optional<TransportProtos.KeyValueProto> getFromSparkplugBMetricToKeyValueProto(String key, SparkplugBProto.Payload.Metric protoMetric) throws ThingsboardException {
+        // Check if the null flag has been set indicating that the value is null
+        if (protoMetric.getIsNull()) {
+            return null;
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:49]: return Optional.empty()

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java [hunk 3]
```diff
@@ -375,14 +375,60 @@ private void handleGatewayPublishMsg(ChannelHandlerContext ctx, String topicName
         }
     }
 
-    private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicName, int msgId, MqttPublishMessage mqttMsg) {
+    private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicName, MqttPublishMessage mqttMsgOld) {
+        MqttPublishMessage mqttMsg = sparkplugSessionHandler.reCreateMqttPublishMessageWithPacketId(mqttMsgOld);
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:379]: No need to repack the message. If the QoS is 0, the msgId is -1

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        }
+        // Otherwise convert the value based on the type
+        int metricType = protoMetric.getDatatype();
+        switch (MetricDataType.fromInteger(metricType)) {
+            case Boolean:
+                return Optional.of(TransportProtos.KeyValueProto.newBuilder().setKey(key).setType(TransportProtos.KeyValueType.BOOLEAN_V)
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:55]: Extract ```TransportProtos.KeyValueProto.newBuilder()``` to the variable to simplify the code.

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                            .setLongV(protoMetric.getIntValue()).build());
+                } else if (protoMetric.hasLongValue()) {
+                    return Optional.of(TransportProtos.KeyValueProto.newBuilder().setKey(key).setType(TransportProtos.KeyValueType.LONG_V)
+                            .setLongV(protoMetric.getLongValue()).build());
+                } else {
+                    log.error("Invalid value for UInt32 datatype");
+                    throw new ThingsboardException("Invalid value for UInt32 datatype " + metricType, ThingsboardErrorCode.INVALID_ARGUMENTS);
+                }
+            case UInt64:
+                return Optional.of(TransportProtos.KeyValueProto.newBuilder().setKey(key).setType(TransportProtos.KeyValueType.DOUBLE_V)
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:90]: Use LONG_V

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+                while (dateTimeByteBuffer.hasRemaining()) {
+                    long longValue = dateTimeByteBuffer.getLong();
+                    dateTimeList.add(longValue);
+                }
+                Gson gson = new GsonBuilder().create();
+                JsonArray dateTimeArray = gson.toJsonTree(dateTimeList).getAsJsonArray();
+                return Optional.of(TransportProtos.KeyValueProto.newBuilder().setKey(key).setType(TransportProtos.KeyValueType.JSON_V)
+                        .setStringV(dateTimeArray.toString()).build());
+            case Unknown:
+            default:
```
Comment (ashvayka) [inline common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:160]: Simply ignore "File", "DataSet', "Template" and "*Array" (return Optional.empty()).

Comment (ashvayka) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:160]: You can ignore File

Comment (ashvayka) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:160]: Also, add to JacksonUtil method ```newArrayNode()```

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java (hunk 2 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:49]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java (hunk 3 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:379]: Ok

Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java:379]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java (hunk 4 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:55]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java (hunk 5 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:90]: Ok

File: common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java (hunk 6 above)
Comment (nickAS21) [reply common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/util/sparkplug/SparkplugMetricUtil.java:158]: Ок

**Pull request merged**